frontend/packages/policy-editor/src/components/PolicyCardRules/PolicyRule/SubResources/ResourceNarrowingList/ResourceNarrowingList.tsx

```diff
@@ -0,0 +1,113 @@
+import React from 'react';
+import classes from './ResourceNarrowingList.module.css';
+import { PolicyResourceFields } from './PolicyResourceFields';
+import { ExpandablePolicyElement } from '../../ExpandablePolicyElement';
+import { StudioButton } from '@studio/components';
+import { PlusIcon } from '@studio/icons';
+import type { PolicyRuleResource } from '../../../../../types';
+import { useTranslation } from 'react-i18next';
+import { usePolicyEditorContext } from '../../../../../contexts/PolicyEditorContext';
+import { usePolicyRuleContext } from '../../../../../contexts/PolicyRuleContext';
+import { getUpdatedRules } from '../../../../../utils/PolicyRuleUtils';
+import { ObjectUtils } from '@studio/pure-functions';
+
+export type ResourceNarrowingListProps = {
+  resources: PolicyRuleResource[];
+  resourceIndex: number;
+};
+
+export const ResourceNarrowingList = ({
+  resources,
+  resourceIndex,
+}: ResourceNarrowingListProps): React.ReactNode => {
+  const { usageType, setPolicyRules, policyRules, savePolicy } = usePolicyEditorContext();
+  const { policyRule, setPolicyError, policyError } = usePolicyRuleContext();
+
+  const { t } = useTranslation();
+
+  const handleDeleteResourceGroup = () => {
+    const updatedResources = [...policyRule.resources];
+    updatedResources.splice(resourceIndex, 1);
+    const updatedRules = getUpdatedRules(
+      { ...policyRule, resources: updatedResources },
+      policyRule.ruleId,
+      policyRules,
+    );
+    setPolicyRules(updatedRules);
+    savePolicy(updatedRules);
+    setPolicyError({ ...policyError, resourceError: updatedResources.length === 0 });
+  };
+
+  const handleCloneResourceGroup = () => {
+    const resourceGroupToDuplicate: PolicyRuleResource[] = policyRule.resources[resourceIndex];
+    const deepCopiedResourceGroupToDuplicate: PolicyRuleResource[] =
+      ObjectUtils.deepCopy(resourceGroupToDuplicate);
+
+    const updatedResources = [...policyRule.resources, deepCopiedResourceGroupToDuplicate];
+    const updatedRules = getUpdatedRules(
+      { ...policyRule, resources: updatedResources },
+      policyRule.ruleId,
+      policyRules,
+    );
+    setPolicyRules(updatedRules);
+    savePolicy(updatedRules);
+  };
+
+  const handleClickAddResourceNarrowing = () => {
+    const newResource: PolicyRuleResource = {
+      type: '',
+      id: '',
+    };
+    const updatedResources = [...policyRule.resources];
+    updatedResources[resourceIndex].push(newResource);
+
+    const updatedRules = getUpdatedRules(
+      { ...policyRule, resources: updatedResources },
+      policyRule.ruleId,
+      policyRules,
+    );
+    setPolicyRules(updatedRules);
+    savePolicy(updatedRules);
+  };
```

Comment:
```suggestion
  const handleDeleteResourceGroup = () => {
    const updatedResources = [...policyRule.resources];
    updatedResources.splice(resourceIndex, 1);
    updatePolicyStates(updatedResources);
    setPolicyError({ ...policyError, resourceError: updatedResources.length === 0 });
  };

  const handleCloneResourceGroup = () => {
    const resourceGroupToDuplicate: PolicyRuleResource[] = policyRule.resources[resourceIndex];
    const deepCopiedResourceGroupToDuplicate: PolicyRuleResource[] =
      ObjectUtils.deepCopy(resourceGroupToDuplicate);
    const updatedResources = [...policyRule.resources, deepCopiedResourceGroupToDuplicate];
    updatePolicyStates(updatedResources);
  };

  const handleClickAddResourceNarrowing = () => {
    const newResource: PolicyRuleResource = {
      type: '',
      id: '',
    };
    const updatedResources = [...policyRule.resources];
    updatedResources[resourceIndex].push(newResource);
    updatePolicyStates(updatedResources);
  };
  
  const updatePolicyStates = (updatedResources): void => {
     const updatedRules = getUpdatedRules(
            { ...policyRule, resources: updatedResources },
            policyRule.ruleId,
            policyRules,
     );
     setPolicyRules(updatedRules);
     savePolicy(updatedRules);
  };
```
What do you think about moving the last logic for each event to a common function (i.e. updatePolicyStates)?

Reply:
I love it, perfect 😄